### ISSUE TITLE
RasterDataset: res does not change when crs changes

### ISSUE THREAD

**Issue by erob-archim** [microsoft/torchgeo#2190]
### Description

It is possible to create an invalid combination of `GridGeoSampler` and `RasterDataset`. This is because `.res` can be in units which do not match the CRS, specifically when the CRS is angular (e.g. EPSG:4326) rather than a planar grid.

It seems like this was somewhat known when pixel space sizes was introduced (https://github.com/microsoft/torchgeo/issues/279#issuecomment-994297653).

I think [these](https://github.com/microsoft/torchgeo/blob/6c2ed0b93b5fd3a22fac7b6933fde0330210e64f/torchgeo/samplers/single.py#L205-L207) are the relevant lines

### Steps to reproduce

```python
import os
import tempfile

from torch.utils.data import DataLoader
from torchgeo.datasets import NAIP, stack_samples
from torchgeo.datasets.utils import download_url
from torchgeo.samplers import GridGeoSampler
from torchgeo.samplers.constants import Units

naip_root = os.path.join(tempfile.gettempdir(), "naip")
naip_url = (
    "https://naipeuwest.blob.core.windows.net/naip/v002/de/2018/de_060cm_2018/38075/"
)
tile = "m_3807511_ne_18_060_20181104.tif"
download_url(naip_url + tile, naip_root)

naip = NAIP(naip_root, crs="EPSG:4326")  # This doesn't work
# naip = NAIP(naip_root)  # This works

sampler = GridGeoSampler(naip, size=(1000, 1000), stride=(1000, 1000), units=Units.PIXELS)
dataloader = DataLoader(naip, sampler=sampler, collate_fn=stack_samples)

first_batch = next(iter(dataloader))
print(first_batch["crs"])
print(first_batch["bbox"])
print(first_batch["image"].shape)
```

### Version

0.5.2

**Comment by adamjstewart** [comment]
Thanks for the code to reproduce, this is extremely useful! Looking into this now. Discoveries I've made so far:

* It isn't specific to GridGeoSampler, RandomGeoSampler also fails
* It works if you use GDAL to reproject manually

**Comment by adamjstewart** [comment]
Current workaround is to specify the resolution yourself:
```python
naip = NAIP(naip_root, crs="EPSG:4326", res=6.052321988618891e-06)
```
Let me see if I can figure out how to compute the new resolution myself.

**Comment by adamjstewart** [comment]
For RasterDataset, I think the ideal situation (depending on user input) would be:

|  |  `crs=CRS` | `crs=None` |
| --- | --- | --- |
| `res=RES` | set both | set res, detect crs |
| `res=None` | set crs, calculate res | detect both |

The only thing we're currently missing is "calculate res", which led to the issue you saw. Note that this does not apply to `VectorDataset` where `res=None` is not allowed. I don't _think_ this matters for `Intersection/UnionDataset` but this is worth thinking about too.

**Comment by adamjstewart** [comment]
Another related bug is that changing the CRS after initialization also doesn't update res (not sure if it should though):
```python
naip = NAIP(naip_root)
print(naip.crs, naip.res)
naip.crs = "EPSG:4326"
print(naip.crs, naip.res)  # crs changed, but not res
```

**Comment by adamjstewart** [comment]
I will probably ignore this related bug until someone complains about it. I don't know off the top of my head how to compute resolution in the new coordinate system, and it seems like it depends on the location, which we know for a single image, but not for the entire R-tree.

**Comment by erob-archim** [comment]
Perhaps there shouldn't be a recalculation in this case, but using `Units.PIXELS` should raise an Exception when the resolution is invalid?

**Comment by adamjstewart** [comment]
I just don't know how to tell if the resolution is invalid.

**Comment by erob-archim** [comment]
Any change from a Cartesian to an angular crs?

**Comment by adamjstewart** [comment]
But what range of resolutions are valid for Cartesian and what range of resolutions are valid for angular? `res` is just a float.

**Comment by adamjstewart** [comment]
Note to self: `crs.coordinate_system.name` only works for pyproj, not rasterio. We currently use rasterio, but we could certainly switch to pyproj. It's already a dependency anyway.

I'm a little worried about `self.res = None`. Things like data loading _should_ work (`rasterio.merge.merge` supports `res=None`) but other stuff like `IntersectionDataset` will probably break. I would be more comfortable with a warning message ("Hey, you might want to change your res too"). It won't be easy to tell if the user is changing crs&res or just crs without some hacking.

I think I'm still leaning towards ignoring this until someone complains, but I could be persuaded.

**Comment by erob-archim** [comment]
Happy with your decision to ignore this, but in case you come back to it some day, the check could be for units in the `rasterio` CRS. They're given in `CRS.to_wkt()`.
e.g.
```python
CRS.from_epsg(32631).to_wkt()
# 'PROJCS["WGS 84 / UTM zone 31N",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Transverse_Mercator"],PARAMETER["latitude_of_origin",0],PARAMETER["central_meridian",3],PARAMETER["scale_factor",0.9996],PARAMETER["false_easting",500000],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],AUTHORITY["EPSG","32631"]]'
# UNIT["meter",1,...]

CRS.from_epsg(4326).to_wkt()
# 'GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AXIS["Latitude",NORTH],AXIS["Longitude",EAST],AUTHORITY["EPSG","4326"]]'
# UNIT["degree", ...]
